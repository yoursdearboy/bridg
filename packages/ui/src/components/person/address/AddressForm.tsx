import {
  Alert,
  Button,
  Group,
  LoadingOverlay,
  Stack,
  TextInput,
} from "@mantine/core";
import { useForm } from "@mantine/form";
import type { UseMutationResult } from "@tanstack/react-query";
import type { PostalAddress, PostalAddressData } from "api-ts";
import { useTranslation } from "react-i18next";

interface AddressFormProps {
  initialValues: PostalAddressData;
  onCancel: () => void;
  mutation: UseMutationResult<PostalAddress, Error, PostalAddressData, unknown>;
}

export const AddressForm = ({
  initialValues,
  onCancel,
  mutation,
}: AddressFormProps) => {
  const { t } = useTranslation();

  const form = useForm<PostalAddressData>({
    initialValues: initialValues,
    validate: {
      street: (value) =>
        value
          ? null
          : t("fieldRequiredMessage", { fieldName: t("Address.street") }),
      country: (value) =>
        value
          ? null
          : t("fieldRequiredMessage", { fieldName: t("Address.country") }),
    },
  });

  const handleSubmit = (data: PostalAddressData) => mutation.mutate(data);

  return (
    <form onSubmit={form.onSubmit(handleSubmit)}>
      <Stack gap="md" pos="relative">
        <LoadingOverlay visible={mutation.isPending} />
        {mutation.error && (
          <Alert color="red" mb="md">
            {mutation.error.message}
          </Alert>
        )}
        <Group grow>
          <TextInput
            label={t("Address.country")}
            withAsterisk
            {...form.getInputProps("country")}
          />
          <TextInput
            label={t("Address.street")}
            withAsterisk
            {...form.getInputProps("street")}
          />
        </Group>

        <details>
          <summary>{t("extra")}</summary>
          <Group grow>
            <TextInput
              label={t("Address.building")}
              {...form.getInputProps("building")}
            />
            <TextInput
              label={t("Address.municipality")}
              {...form.getInputProps("municipality")}
            />
            <TextInput
              label={t("Address.state")}
              {...form.getInputProps("state")}
            />
          </Group>
          <TextInput label={t("Address.zip")} {...form.getInputProps("zip")} />
        </details>

        <Group justify="flex-end" mt="md">
          <Button
            variant="outline"
            onClick={onCancel}
            disabled={mutation.isPending}
          >
            {t("cancel")}
          </Button>
          <Button type="submit" loading={mutation.isPending}>
            {t("submit")}
          </Button>
        </Group>
      </Stack>
    </form>
  );
};
