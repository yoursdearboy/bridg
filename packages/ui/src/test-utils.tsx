/* eslint-disable react-refresh/only-export-components */
import { MantineProvider } from "@mantine/core";
import { DatesProvider } from "@mantine/dates";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import {
  type AnyRoute,
  createRootRoute,
  createRootRouteWithContext,
  createRoute,
  createRouter,
  RouterProvider,
} from "@tanstack/react-router";
import { act, render } from "@testing-library/react";
import "./i18n";
import type { PropsWithChildren } from "react";

const queryClient = new QueryClient();

const App = ({ children }: React.PropsWithChildren) => {
  return (
    <QueryClientProvider client={queryClient}>
      <MantineProvider defaultColorScheme="auto">
        <DatesProvider settings={{ locale: "en" }}>{children}</DatesProvider>
      </MantineProvider>
    </QueryClientProvider>
  );
};

export const renderRoute = (
  r: AnyRoute,
  { params, props }: { params?: object; props?: PropsWithChildren } = {}
) => {
  const rootRoute = createRootRouteWithContext()();
  let routeConfig = {
    ...r.options,
    params: params,
    getParentRoute: () => rootRoute,
    path: "/",
  };
  if (props?.children)
    routeConfig = { ...routeConfig, component: () => props.children };

  const testRoute = createRoute(routeConfig);
  const routeTree = rootRoute.addChildren([testRoute]);
  const router = createRouter({
    routeTree,
    context: {
      queryClient,
    },
  });
  return act(() =>
    render(
      <App>
        <RouterProvider router={router} />
      </App>
    )
  );
};

export const renderComponent = (component: React.ReactNode) =>
  render(<App>{component}</App>);

export const EmptyRouterProvider = (props: PropsWithChildren) => {
  const rootRoute = createRootRoute({
    component: () => props.children,
  });

  return renderRoute(rootRoute, { props: { children: props.children } });
};
