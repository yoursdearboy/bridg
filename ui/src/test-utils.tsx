import { MantineProvider } from "@mantine/core";
import { DatesProvider } from "@mantine/dates";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import {
  createRootRoute,
  createRoute,
  createRouter,
  RouterProvider,
  type AnyRoute,
} from "@tanstack/react-router";
import { act, render } from "@testing-library/react";
import "./i18n";

// eslint-disable-next-line react-refresh/only-export-components
const App = ({ children }: React.PropsWithChildren) => {
  const queryClient = new QueryClient();

  return (
    <QueryClientProvider client={queryClient}>
      <MantineProvider defaultColorScheme="auto">
        <DatesProvider settings={{ locale: "en" }}>{children}</DatesProvider>
      </MantineProvider>
    </QueryClientProvider>
  );
};

export const renderRoute = (
  r: AnyRoute,
  { params }: { params?: object } = {}
) => {
  const rootRoute = createRootRoute();
  const testRoute = createRoute({
    ...r.options,
    params: params,
    getParentRoute: () => rootRoute,
    path: "/",
  });
  const routeTree = rootRoute.addChildren([testRoute]);
  const router = createRouter({
    routeTree,
  });
  return act(() =>
    render(
      <App>
        <RouterProvider router={router} />
      </App>
    )
  );
};
