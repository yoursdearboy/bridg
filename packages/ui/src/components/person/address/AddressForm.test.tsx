import { renderComponent } from "@/test-utils";
import type { PostalAddress, PostalAddressData } from "api-ts";
import { describe, expect, it, vi } from "vitest";
import { AddressForm } from "./AddressForm";
import { type UseMutationResult } from "@tanstack/react-query";

describe("AddressForm", () => {
  it("matches snapshot", () => {
    const initialValues: PostalAddressData = {
      street: "Red",
      building: "Square",
      use: "h",
    };
    const result = {
      id: "1",
      label: "Red Square",
      ...initialValues,
    };
    const mockMutation: UseMutationResult<
      PostalAddress,
      Error,
      PostalAddressData,
      unknown
    > = {
      data: result,
      error: null,
      isError: false,
      isIdle: false,
      isPending: false,
      isPaused: false,
      isSuccess: true,
      failureCount: 0,
      failureReason: null,
      mutate: vi.fn(),
      mutateAsync: vi.fn(),
      reset: vi.fn(),
      submittedAt: 0,
      variables: {},
      status: "success",
      context: {},
    };

    const { asFragment } = renderComponent(
      <AddressForm
        mutation={mockMutation}
        initialValues={initialValues}
        onCancel={() => {}}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
