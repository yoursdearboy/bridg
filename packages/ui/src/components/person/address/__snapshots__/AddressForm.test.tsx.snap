// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddressForm > matches snapshot 1`] = `
<DocumentFragment>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <form>
    <div
      class="m_6d731127 mantine-Stack-root"
      style="--stack-gap: var(--mantine-spacing-md); --stack-align: stretch; --stack-justify: flex-start; position: relative;"
    >
      <div
        class="m_4081bf90 mantine-Group-root"
        data-grow="true"
        style="--group-child-width: calc(50% - (var(--mantine-spacing-md) - var(--mantine-spacing-md) / 2)); --group-gap: var(--mantine-spacing-md); --group-align: center; --group-justify: flex-start; --group-wrap: wrap;"
      >
        <div
          class="m_46b77525 mantine-InputWrapper-root mantine-TextInput-root"
        >
          <label
            class="m_8fdc1311 mantine-InputWrapper-label mantine-TextInput-label"
            data-required="true"
            for="mantine-4miilw5eg"
            id="mantine-4miilw5eg-label"
          >
            Country
            <span
              aria-hidden="true"
              class="m_78a94662 mantine-InputWrapper-required mantine-TextInput-required"
            >
               *
            </span>
          </label>
          <div
            class="m_6c018570 mantine-Input-wrapper mantine-TextInput-wrapper"
            data-variant="default"
          >
            <input
              aria-invalid="false"
              class="m_8fb7ebe7 mantine-Input-input mantine-TextInput-input"
              data-path="country"
              data-variant="default"
              id="mantine-4miilw5eg"
              value=""
            />
          </div>
        </div>
        <div
          class="m_46b77525 mantine-InputWrapper-root mantine-TextInput-root"
        >
          <label
            class="m_8fdc1311 mantine-InputWrapper-label mantine-TextInput-label"
            data-required="true"
            for="mantine-17gjn16kc"
            id="mantine-17gjn16kc-label"
          >
            Street
            <span
              aria-hidden="true"
              class="m_78a94662 mantine-InputWrapper-required mantine-TextInput-required"
            >
               *
            </span>
          </label>
          <div
            class="m_6c018570 mantine-Input-wrapper mantine-TextInput-wrapper"
            data-variant="default"
          >
            <input
              aria-invalid="false"
              class="m_8fb7ebe7 mantine-Input-input mantine-TextInput-input"
              data-path="street"
              data-variant="default"
              id="mantine-17gjn16kc"
              value="Red"
            />
          </div>
        </div>
      </div>
      <details>
        <summary>
          Extra
        </summary>
        <div
          class="m_4081bf90 mantine-Group-root"
          data-grow="true"
          style="--group-child-width: calc(33.333333333333336% - (var(--mantine-spacing-md) - var(--mantine-spacing-md) / 3)); --group-gap: var(--mantine-spacing-md); --group-align: center; --group-justify: flex-start; --group-wrap: wrap;"
        >
          <div
            class="m_46b77525 mantine-InputWrapper-root mantine-TextInput-root"
          >
            <label
              class="m_8fdc1311 mantine-InputWrapper-label mantine-TextInput-label"
              for="mantine-2pbrvloro"
              id="mantine-2pbrvloro-label"
            >
              Building
            </label>
            <div
              class="m_6c018570 mantine-Input-wrapper mantine-TextInput-wrapper"
              data-variant="default"
            >
              <input
                aria-invalid="false"
                class="m_8fb7ebe7 mantine-Input-input mantine-TextInput-input"
                data-path="building"
                data-variant="default"
                id="mantine-2pbrvloro"
                value="Square"
              />
            </div>
          </div>
          <div
            class="m_46b77525 mantine-InputWrapper-root mantine-TextInput-root"
          >
            <label
              class="m_8fdc1311 mantine-InputWrapper-label mantine-TextInput-label"
              for="mantine-pfqyzivqd"
              id="mantine-pfqyzivqd-label"
            >
              Municipality
            </label>
            <div
              class="m_6c018570 mantine-Input-wrapper mantine-TextInput-wrapper"
              data-variant="default"
            >
              <input
                aria-invalid="false"
                class="m_8fb7ebe7 mantine-Input-input mantine-TextInput-input"
                data-path="municipality"
                data-variant="default"
                id="mantine-pfqyzivqd"
                value=""
              />
            </div>
          </div>
          <div
            class="m_46b77525 mantine-InputWrapper-root mantine-TextInput-root"
          >
            <label
              class="m_8fdc1311 mantine-InputWrapper-label mantine-TextInput-label"
              for="mantine-7lzvq3tar"
              id="mantine-7lzvq3tar-label"
            >
              State
            </label>
            <div
              class="m_6c018570 mantine-Input-wrapper mantine-TextInput-wrapper"
              data-variant="default"
            >
              <input
                aria-invalid="false"
                class="m_8fb7ebe7 mantine-Input-input mantine-TextInput-input"
                data-path="state"
                data-variant="default"
                id="mantine-7lzvq3tar"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="m_46b77525 mantine-InputWrapper-root mantine-TextInput-root"
        >
          <label
            class="m_8fdc1311 mantine-InputWrapper-label mantine-TextInput-label"
            for="mantine-m75qc1lk3"
            id="mantine-m75qc1lk3-label"
          >
            Zip
          </label>
          <div
            class="m_6c018570 mantine-Input-wrapper mantine-TextInput-wrapper"
            data-variant="default"
          >
            <input
              aria-invalid="false"
              class="m_8fb7ebe7 mantine-Input-input mantine-TextInput-input"
              data-path="zip"
              data-variant="default"
              id="mantine-m75qc1lk3"
              value=""
            />
          </div>
        </div>
      </details>
      <div
        class="m_4081bf90 mantine-Group-root"
        style="--group-gap: var(--mantine-spacing-md); --group-align: center; --group-justify: flex-end; --group-wrap: wrap; margin-top: var(--mantine-spacing-md);"
      >
        <button
          class="mantine-focus-auto mantine-active m_77c9d27d mantine-Button-root m_87cf2631 mantine-UnstyledButton-root"
          data-variant="outline"
          style="--button-bg: transparent; --button-hover: var(--mantine-color-blue-outline-hover); --button-color: var(--mantine-color-blue-outline); --button-bd: calc(0.0625rem * var(--mantine-scale)) solid var(--mantine-color-blue-outline);"
          type="button"
        >
          <span
            class="m_80f1301b mantine-Button-inner"
          >
            <span
              class="m_811560b9 mantine-Button-label"
            >
              Cancel
            </span>
          </span>
        </button>
        <button
          class="mantine-focus-auto mantine-active m_77c9d27d mantine-Button-root m_87cf2631 mantine-UnstyledButton-root"
          style="--button-color: var(--mantine-color-white);"
          type="submit"
        >
          <span
            class="m_80f1301b mantine-Button-inner"
          >
            <span
              class="m_811560b9 mantine-Button-label"
            >
              Submit
            </span>
          </span>
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
